system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,47 +6,80 @@
 // Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
+#include "file.h"
+#include "file_table.h"
+#include "open_file_descriptor.h"
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <errno.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+__wasi_fd_t __syscall63(__wasi_fd_t oldfd, __wasi_fd_t newfd) { // dup2
+  FileTable::Handle currentHandle = FileTable::get();
+
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (oldfd >= currentHandle.getSize() || oldfd < 0) {
+    return (__wasi_fd_t)(-1);
+  }
+
+  std::shared_ptr<OpenFileDescriptor> oldOpenFile = currentHandle[oldfd];
+
+  if (oldOpenFile == nullptr) {
+    return (__wasi_fd_t)(-1);
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  // If the file descriptor newfd was previously open, it is closed
+  // before being reused; the close is performed silently.
+  if (newfd < 0) {
+    return (__wasi_fd_t)(-1);
+  }
+
+  if (newfd < currentHandle.getSize()) {
+    std::shared_ptr<OpenFileDescriptor> newOpenFile = currentHandle[newfd];
+    if (newOpenFile) {
```

Comment:
Another improvement you can make to the API here would be to define `operator bool()` for `Entry` that returns whether the entry exists and contains an open file. Then this code could just be:
```suggestion
  if (currentHandle[newfd]) {
```